File: backend/src/settings/settings.service.ts
import { Injectable, OnModuleInit } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Setting } from './entities/setting.entity';
import { UpdateAuthSettingsDto } from './dto/update-auth-settings.dto';
import { UpdateEmailSettingsDto } from './dto/update-email-settings.dto';
import { UpdateAlertSettingsDto } from './dto/update-alert-settings.dto';

@Injectable()
export class SettingsService implements OnModuleInit {
  constructor(
    @InjectRepository(Setting)
    private settingsRepository: Repository<Setting>,
  ) {}

  async onModuleInit() {
    await this.initializeSettings();
  }

  private async initializeSettings() {
    const defaultSettings = [
      { key: 'AUTH_GITHUB_ENABLED', value: 'false' },
      { key: 'AUTH_KEYCLOAK_ENABLED', value: 'false' },
      { key: 'SMTP_HOST', value: '' },
      { key: 'SMTP_PORT', value: '' },
      { key: 'SMTP_USER', value: '' },
      { key: 'SMTP_PASS', value: '' },
      { key: 'SMTP_FROM', value: '' },
      { key: 'ALERT_EMAILS', value: '' },
      { key: 'ALERT_MIN_SEVERITY', value: 'HIGH' },
      { key: 'TEAMS_WEBHOOK_URL', value: '' },
      { key: 'TEAMS_ENABLED', value: 'false' },
    ];

    for (const setting of defaultSettings) {
      const exists = await this.settingsRepository.findOne({ where: { key: setting.key } });
      if (!exists) {
        await this.settingsRepository.save(setting);
      }
    }
  }

  // --- Auth Settings ---
  async getAuthSettings() {
    const githubSetting = await this.settingsRepository.findOne({ where: { key: 'AUTH_GITHUB_ENABLED' } });
    const keycloakSetting = await this.settingsRepository.findOne({ where: { key: 'AUTH_KEYCLOAK_ENABLED' } });

    return {
      githubEnabled: githubSetting?.value === 'true',
      keycloakEnabled: keycloakSetting?.value === 'true',
    };
  }

  async updateAuthSettings(dto: UpdateAuthSettingsDto) {
    if (dto.githubEnabled !== undefined) {
      await this.settingsRepository.save({ key: 'AUTH_GITHUB_ENABLED', value: dto.githubEnabled.toString() });
    }
    if (dto.keycloakEnabled !== undefined) {
      await this.settingsRepository.save({ key: 'AUTH_KEYCLOAK_ENABLED', value: dto.keycloakEnabled.toString() });
    }
    return this.getAuthSettings();
  }

  // --- Email Settings ---
  async getEmailSettings() {
    const host = await this.settingsRepository.findOne({ where: { key: 'SMTP_HOST' } });
    const port = await this.settingsRepository.findOne({ where: { key: 'SMTP_PORT' } });
    const user = await this.settingsRepository.findOne({ where: { key: 'SMTP_USER' } });
    const pass = await this.settingsRepository.findOne({ where: { key: 'SMTP_PASS' } });
    const from = await this.settingsRepository.findOne({ where: { key: 'SMTP_FROM' } });

    return {
      smtpHost: host?.value || '',
      smtpPort: port?.value || '',
      smtpUser: user?.value || '',
      smtpPass: pass?.value || '',
      smtpFrom: from?.value || '',
    };
  }

  async updateEmailSettings(dto: UpdateEmailSettingsDto) {
    if (dto.smtpHost !== undefined) await this.settingsRepository.save({ key: 'SMTP_HOST', value: dto.smtpHost });
    if (dto.smtpPort !== undefined) await this.settingsRepository.save({ key: 'SMTP_PORT', value: dto.smtpPort });
    if (dto.smtpUser !== undefined) await this.settingsRepository.save({ key: 'SMTP_USER', value: dto.smtpUser });
    if (dto.smtpPass !== undefined) await this.settingsRepository.save({ key: 'SMTP_PASS', value: dto.smtpPass });
    if (dto.smtpFrom !== undefined) await this.settingsRepository.save({ key: 'SMTP_FROM', value: dto.smtpFrom });
    return this.getEmailSettings();
  }

  // --- Alert Settings ---
  async getAlertSettings() {
    const emails = await this.settingsRepository.findOne({ where: { key: 'ALERT_EMAILS' } });
    const minSeverity = await this.settingsRepository.findOne({ where: { key: 'ALERT_MIN_SEVERITY' } });

    return {
      alertEmails: emails?.value || '',
      alertMinSeverity: minSeverity?.value || 'HIGH',
    };
  }

  async updateAlertSettings(dto: UpdateAlertSettingsDto) {
    if (dto.alertEmails !== undefined) await this.settingsRepository.save({ key: 'ALERT_EMAILS', value: dto.alertEmails });
    if (dto.alertMinSeverity !== undefined) await this.settingsRepository.save({ key: 'ALERT_MIN_SEVERITY', value: dto.alertMinSeverity });
    return this.getAlertSettings();
  }

  // --- Teams Settings ---
  async getTeamsSettings() {
    const url = await this.settingsRepository.findOne({ where: { key: 'TEAMS_WEBHOOK_URL' } });
    const enabled = await this.settingsRepository.findOne({ where: { key: 'TEAMS_ENABLED' } });

    return {
      webhookUrl: url?.value || '',
      enabled: enabled?.value === 'true',
    };
  }

  async updateTeamsSettings(dto: { webhookUrl?: string, enabled?: boolean }) {
    if (dto.webhookUrl !== undefined) await this.settingsRepository.save({ key: 'TEAMS_WEBHOOK_URL', value: dto.webhookUrl });
    if (dto.enabled !== undefined) await this.settingsRepository.save({ key: 'TEAMS_ENABLED', value: dto.enabled.toString() });
    return this.getTeamsSettings();
  }
}

File: backend-java/src/main/java/com/asmolabs/zanshin/settings/services/SettingsService.java
package com.asmolabs.zanshin.settings.services;

import com.asmolabs.zanshin.repository.services.AuditLogService;
import com.asmolabs.zanshin.settings.dto.UpdateAlertSettingsDto;
import com.asmolabs.zanshin.settings.dto.UpdateAuthSettingsDto;
import com.asmolabs.zanshin.settings.dto.UpdateEmailSettingsDto;
import com.asmolabs.zanshin.settings.entities.Setting;
import com.asmolabs.zanshin.settings.repositories.SettingRepository;
import jakarta.annotation.PostConstruct;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class SettingsService {

    private final SettingRepository settingRepository;
    private final AuditLogService auditService;

    @PostConstruct
    public void init() {
        initializeSettings();
    }

    private void initializeSettings() {
        Map<String, String> defaultSettings = new HashMap<>();
        defaultSettings.put("AUTH_GITHUB_ENABLED", "false");
        defaultSettings.put("AUTH_KEYCLOAK_ENABLED", "false");
        defaultSettings.put("SMTP_HOST", "");
        defaultSettings.put("SMTP_PORT", "");
        defaultSettings.put("SMTP_USER", "");
        defaultSettings.put("SMTP_PASS", "");
        defaultSettings.put("SMTP_FROM", "");
        defaultSettings.put("ALERT_EMAILS", "");
        defaultSettings.put("ALERT_MIN_SEVERITY", "HIGH");
        defaultSettings.put("TEAMS_WEBHOOK_URL", "");
        defaultSettings.put("TEAMS_ENABLED", "false");

        for (Map.Entry<String, String> entry : defaultSettings.entrySet()) {
            if (!settingRepository.existsById(entry.getKey())) {
                settingRepository.save(new Setting(entry.getKey(), entry.getValue()));
            }
        }
    }

    // --- Auth Settings ---
    public Map<String, Boolean> getAuthSettings() {
        boolean githubEnabled = Boolean.parseBoolean(settingRepository.findById("AUTH_GITHUB_ENABLED").map(Setting::getValue).orElse("false"));
        boolean keycloakEnabled = Boolean.parseBoolean(settingRepository.findById("AUTH_KEYCLOAK_ENABLED").map(Setting::getValue).orElse("false"));

        return Map.of(
            "githubEnabled", githubEnabled,
            "keycloakEnabled", keycloakEnabled
        );
    }

    @Transactional
    public Map<String, Boolean> updateAuthSettings(UpdateAuthSettingsDto dto, String userId) {
        if (dto.githubEnabled() != null) {
            settingRepository.save(new Setting("AUTH_GITHUB_ENABLED", dto.githubEnabled().toString()));
        }
        if (dto.keycloakEnabled() != null) {
            settingRepository.save(new Setting("AUTH_KEYCLOAK_ENABLED", dto.keycloakEnabled().toString()));
        }

        auditService.logAction(userId, "SETTINGS", "UPDATE", "Auth settings updated.");
        return getAuthSettings();
    }

    // --- Email Settings ---
    public Map<String, String> getEmailSettings() {
        return Map.of(
            "smtpHost", settingRepository.findById("SMTP_HOST").map(Setting::getValue).orElse(""),
            "smtpPort", settingRepository.findById("SMTP_PORT").map(Setting::getValue).orElse(""),
            "smtpUser", settingRepository.findById("SMTP_USER").map(Setting::getValue).orElse(""),
            "smtpPass", settingRepository.findById("SMTP_PASS").map(Setting::getValue).orElse(""),
            "smtpFrom", settingRepository.findById("SMTP_FROM").map(Setting::getValue).orElse("")
        );
    }

    @Transactional
    public Map<String, String> updateEmailSettings(UpdateEmailSettingsDto dto, String userId) {
        if (dto.smtpHost() != null) settingRepository.save(new Setting("SMTP_HOST", dto.smtpHost()));
        if (dto.smtpPort() != null) settingRepository.save(new Setting("SMTP_PORT", dto.smtpPort()));
        if (dto.smtpUser() != null) settingRepository.save(new Setting("SMTP_USER", dto.smtpUser()));
        if (dto.smtpPass() != null) settingRepository.save(new Setting("SMTP_PASS", dto.smtpPass()));
        if (dto.smtpFrom() != null) settingRepository.save(new Setting("SMTP_FROM", dto.smtpFrom()));

        auditService.logAction(userId, "SETTINGS", "UPDATE", "Email settings updated.");
        return getEmailSettings();
    }

    // --- Alert Settings ---
    public Map<String, String> getAlertSettings() {
        return Map.of(
            "alertEmails", settingRepository.findById("ALERT_EMAILS").map(Setting::getValue).orElse(""),
            "alertMinSeverity", settingRepository.findById("ALERT_MIN_SEVERITY").map(Setting::getValue).orElse("HIGH")
        );
    }

    @Transactional
    public Map<String, String> updateAlertSettings(UpdateAlertSettingsDto dto, String userId) {
        if (dto.alertEmails() != null) settingRepository.save(new Setting("ALERT_EMAILS", dto.alertEmails()));
        if (dto.alertMinSeverity() != null) settingRepository.save(new Setting("ALERT_MIN_SEVERITY", dto.alertMinSeverity()));
        
        auditService.logAction(userId, "SETTINGS", "UPDATE", "Alert settings updated.");
        return getAlertSettings();
    }

    // --- Teams Settings ---
    public Map<String, Object> getTeamsSettings() {
        String url = settingRepository.findById("TEAMS_WEBHOOK_URL").map(Setting::getValue).orElse("");
        boolean enabled = Boolean.parseBoolean(settingRepository.findById("TEAMS_ENABLED").map(Setting::getValue).orElse("false"));

        return Map.of(
            "webhookUrl", url,
            "enabled", enabled
        );
    }

    @Transactional
    public Map<String, Object> updateTeamsSettings(Map<String, Object> dto, String userId) {
        if (dto.containsKey("webhookUrl")) {
            settingRepository.save(new Setting("TEAMS_WEBHOOK_URL", (String) dto.get("webhookUrl")));
        }
        if (dto.containsKey("enabled")) {
            settingRepository.save(new Setting("TEAMS_ENABLED", dto.get("enabled").toString()));
        }

        auditService.logAction(userId, "SETTINGS", "UPDATE", "Teams settings updated.");
        return getTeamsSettings();
    }
}
